### PR TITLE
Use compute_loss in prediction_step

### PR DESCRIPTION
**Pull request by sgugger**
# What does this PR do?

As requested in #9915, this PR uses `compute_loss` in the `prediction_step` method of `Trainer`, so it properly computes losses when the user have customized the way to do that. It does require a new argument to `compute_loss` to return the outputs on top of the loss for the prediction loop, so users that want to use this feature will have to tweak their subclass a little bit, but there is no breaking change.

### CODE REVIEW

File: src/transformers/trainer.py [hunk 1]
```diff
@@ -1718,29 +1720,27 @@ def prediction_step(
...
                 if isinstance(outputs, dict):
                     logits = tuple(v for k, v in outputs.items() if k not in ignore_keys)
                 else:
                     logits = outputs
-            # TODO: this needs to be fixed and made cleaner later.
-            if self.args.past_index >= 0:
-                self._past = outputs[self.args.past_index if has_labels else self.args.past_index - 1]
+                # TODO: this needs to be fixed and made cleaner later.
+                if self.args.past_index >= 0:
+                    self._past = outputs[self.args.past_index - 1]
```
Comment (LysandreJik) [inline src/transformers/trainer.py:1743]: This doesn't apply `self._past = outputs[self.args.past_index]` when `has_labels == True`, is that voluntary?

**Review by LysandreJik — APPROVED** [review]
LGTM!

File: src/transformers/trainer.py (hunk 1 above)
Comment (sgugger) [reply src/transformers/trainer.py:1743]: We are in the else branch of `if has_labels` so this does not apply ;-)
For the True case, this is done in `compute_loss` already.
